### PR TITLE
Avoid overwriting infrastructure roles.

### PR DESCRIPTION
**Pull request by alexeyklyukin**
For a role defined in both the infrastructure roles secret and the cluster manifest use the infrastructure role definition and add flags defined in the manifest. #166

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -645,6 +653,10 @@ func (c *Cluster) initHumanUsers() error {
 			}
 		}
 
+		if _, present := c.pgUsers[username]; present {
+			c.logger.Warnf("overwriting existing user %q with the data from the teams API")
```
Comment (zerg-junior) [inline pkg/cluster/cluster.go]: 1. Should the username be passed to the Warnf call as a value ?
2. Is the data from the teams API the only thing that can trigger the overwrite ?

**Comment by zerg-junior** [comment]
@alexeyklyukin Could you please add a small unit test for this case to avoid regressions in the future ?

File: pkg/cluster/cluster.go (hunk 1 above)
Comment (alexeyklyukin) [reply pkg/cluster/cluster.go]: good catch for the username, teams API roles will overwrite roles with the same name obtained from the infrastructure roles or the cluster manifests, but cluster manifest roles will merge with the infrastructure one (same PR), so teams API roles are the only ones that can overwrite other roles.

**Comment by alexeyklyukin** [comment]
Added tests for the initHumanUser and initRobotUsers

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -610,18 +612,27 @@ func (c *Cluster) initSystemUsers() {
 func (c *Cluster) initRobotUsers() error {
 	for username, userFlags := range c.Spec.Users {
 		if !isValidUsername(username) {
-			return fmt.Errorf("invalid username: '%v'", username)
+			return fmt.Errorf("invalid username: %q", username)
```
Comment (zerg-junior) [inline pkg/cluster/cluster.go]: Shall we briefly mention here how valid usernames look like ?

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -610,18 +612,27 @@ func (c *Cluster) initSystemUsers() {
...
+			c.pgUsers[username] = spec.PgUser{
+				Name:     username,
+				Password: util.RandomPassword(constants.PasswordLength),
+				Flags:    flags,
+			}
+		} else {
+			// avoid overwriting the password if the user is already there. The flags should be
+			// merged here, but since there is no mechanism to define them for non-robot roles
+			// they are assigned from the robot user.
+			c.logger.Debugf("merging user %q data", username)
```
Comment (zerg-junior) [inline pkg/cluster/cluster.go]: Shall this message also say in which places exactly user is defined twice ?

**Comment by zerg-junior** [comment]
:+1:

File: pkg/cluster/cluster.go (hunk 3 above)
Comment (alexeyklyukin) [reply pkg/cluster/cluster.go]: good point.

File: pkg/cluster/cluster.go (hunk 2 above)
Comment (alexeyklyukin) [reply pkg/cluster/cluster.go]: It would be too long to fit into the error message, I think it's a task of the documentation.

**Comment by alexeyklyukin** [comment]
👍

**Pull request merged**